src/general_listener.c

```diff
@@ -129,103 +129,191 @@ static void
+        }
+        thread_data.child_time[thread_data.level] = 0.0;
+        thread_data.level++;
+        if (thread_data.level == thread_data.capacity) {
+            thread_data.capacity *= 2;
+            pthread_pause_disable();
+            thread_data.child_time = g_renew(double, thread_data.child_time, thread_data.capacity);
+            pthread_pause_enable();
+        }
+        self->num_calls[index]++;
+        if (peak_detach_count && self->num_calls[index] >= peak_detach_count) {
+            pthread_mutex_lock(&lock);
+            size_t hook_id = self->hook_id;
+            if (!array_listener_detached[hook_id]) {
+                array_listener_detached[hook_id] = TRUE;
+                GumMetalHashTableIter peak_tid_iter;
+                pthread_t peak_tid_key;
+                gum_metal_hash_table_iter_init(&peak_tid_iter, peak_tid_mapping);
+                while (gum_metal_hash_table_iter_next(&peak_tid_iter, (void**)&peak_tid_key, NULL)) {
+                    // g_print ("peak_tid_key %lu my_tid %lu\n", peak_tid_key, my_tid);
+                    if (peak_tid_key != my_tid)
+                        pthread_pause(peak_tid_key);
+                }
+                gum_interceptor_begin_transaction(interceptor);
+                gum_interceptor_detach(interceptor, listener);
+                gum_interceptor_end_transaction(interceptor);
+                gum_metal_hash_table_iter_init(&peak_tid_iter, peak_tid_mapping);
+                while (gum_metal_hash_table_iter_next(&peak_tid_iter, (void**)&peak_tid_key, NULL)) {
+                    if (peak_tid_key != my_tid)
+                        pthread_unpause(peak_tid_key);
+                }
             }
+            pthread_mutex_unlock(&lock);
+            // gum_interceptor_revert(interceptor, hook_address[hook_id]);
+            // g_printerr ("revert hook_id %lu %p\n", hook_id, hook_address[hook_id]);
         }
-        pthread_mutex_unlock(&lock);
-        // gum_interceptor_revert(interceptor, hook_address[hook_id]);
-        // g_printerr ("revert hook_id %lu %p\n", hook_id, hook_address[hook_id]);
+        double* current_time = GUM_IC_GET_INVOCATION_DATA(ic, double);
+        pthread_pause_disable();
+        *current_time = peak_second();
+        // g_printerr ("hook_id %lu time %f count %lu\n", hook_id, *current_time, self->num_calls[mapped_tid]);
     }
-    double* current_time = GUM_IC_GET_INVOCATION_DATA(ic, double);
-    pthread_pause_disable();
-    *current_time = peak_second();
-    // g_printerr ("hook_id %lu time %f count %lu\n", hook_id, *current_time, self->num_calls[mapped_tid]);
 }
 
 static void
 peak_general_listener_on_leave(GumInvocationListener* listener,
                                GumInvocationContext* ic)
 {
-    double end_time = peak_second();
-    pthread_pause_enable();
-    if (!listener || g_object_is_floating(listener)) {
+    if (peak_detach_count == 0) {
+        double end_time = peak_second();
```

Comment:
This line that takes the timing should always be the first to be called to get an accurate time measurement. So, it should be moved out of the if..else.. statement too.